File: app/src/main/java/top/someones/cardmatch/core/BaseGameObserver.java
package top.someones.cardmatch.core;

import android.content.Context;
import android.view.View;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public abstract class BaseGameObserver implements GameObserver {

    private final GameCallback callback;
    private final Context mContext;
    private final int[] mData = new int[16];
    private int mGameProgress = 0;

    public BaseGameObserver(Context context, GameCallback callback) {
        this.mContext = context;
        this.callback = callback;
    }

    private int getData(int index) {
        return mData[index];
    }

    @Override
    public View[][] newGame() {
        System.arraycopy(setData(), 0, mData, 0, mData.length);
        return makeGameView(mData);
    }

    protected abstract int[] setData();

    protected abstract View[][] makeGameView(int[] gameData);

    protected Context getContext() {
        return mContext;
    }

    @Override
    public void check(int a, int b) {
        if (getData(a) == getData(b)) {
            mGameProgress++;
            callback.onSuccess(a, b);
            if (mGameProgress == GameObserver.MAX_VIEW / 2)
                callback.onWin();
        } else
            callback.onFailure(a, b);
    }

    protected int[] getRandomResourcesIndex(int resLength) {
        List<Integer> one = new ArrayList<>(resLength);
        for (int i = 0; i < resLength; i++) {
            one.add(i);
        }
        List<Integer> two = new ArrayList<>(GameObserver.MAX_VIEW);
        Random random = new Random();
        int index;
        for (int i = 0; i < GameObserver.MAX_VIEW / 2; i++) {
            index = random.nextInt(resLength - i);
            two.add(one.get(index));
            one.remove(index);
        }
        two.addAll(two);
        int[] data = new int[GameObserver.MAX_VIEW];
        for (int i = 0; i < GameObserver.MAX_VIEW; i++) {
            index = random.nextInt(GameObserver.MAX_VIEW - i);
            data[i] = two.remove(index);
        }
        return data;
    }

}

File: app/src/main/java/top/someones/cardmatch/core/BaseGameObserver.kt
package top.someones.cardmatch.core

import android.content.Context
import android.view.View
import java.util.*

abstract class BaseGameObserver(
    protected val context: Context,
    private val callback: GameCallback
) : GameObserver {
    private val mData = IntArray(16)
    private var mGameProgress = 0
    private lateinit var mGameToken: String
    private fun getData(index: Int): Int {
        return mData[index]
    }

    override fun newGame(token: String): Array<Array<View?>> {
        mGameToken = token
        mGameProgress = 0
        System.arraycopy(setData()!!, 0, mData, 0, mData.size)
        return makeGameView(mData)
    }

    protected abstract fun setData(): IntArray?
    protected abstract fun makeGameView(gameData: IntArray): Array<Array<View?>>
    override fun check(a: Int, b: Int) {
        if (getData(a) == getData(b)) {
            mGameProgress++
            callback.onSuccess(mGameToken, a, b)
            if (mGameProgress == GameObserver.MAX_VIEW / 2) callback.onWin(mGameToken)
        } else callback.onFailure(mGameToken, a, b)
    }

    protected fun getRandomResourcesIndex(resLength: Int): IntArray {
        val one: MutableList<Int> = ArrayList(resLength)
        for (i in 0 until resLength) {
            one.add(i)
        }
        val two: MutableList<Int> = ArrayList<Int>(GameObserver.MAX_VIEW)
        val random = Random()
        var index: Int
        for (i in 0 until GameObserver.MAX_VIEW / 2) {
            index = random.nextInt(resLength - i)
            two.add(one[index])
            one.removeAt(index)
        }
        two.addAll(two)
        val data = IntArray(GameObserver.MAX_VIEW)
        for (i in 0 until GameObserver.MAX_VIEW) {
            index = random.nextInt(GameObserver.MAX_VIEW - i)
            data[i] = two.removeAt(index)
        }
        return data
    }
}